src/lib/storage/run_length_segment/run_length_segment_iterable.hpp

```diff
@@ -44,143 +43,189 @@ class RunLengthSegmentIterable : public PointAccessibleSegmentIterable<RunLength
  private:
   const RunLengthSegment<T>& _segment;
 
+  /**
+   * Due to the nature of Run Length encoding, point-access is not in O(1).
+   * Run length segments store the end positions of runs in a sorted vector. To access a particular position, this
+   * vector has to be searched for the end position of the according run. It is possible to find the value of a
+   * position in O(log(n)) by doing a binary search.
+   * In case we hav already searched for a nearby position (e.g., when a previous not selective filter removed only
+   * relatively few lines), we can linearly search the end position. Since position lists are often ordered and modern
+   * CPUs are well at prefetching, linear searches are often faster than a complete binary search.
+   * determine_linear_search_offset_distance_threshold() estimates the threshold of when to use a linear or a binary
+   * search. Given the previous and the current chunk offset, the threshold is used to determine if a linear search is
+   * faster for the given number of skipped positions.
+   */
+
+  // LINEAR_SEARCH_ELEMENTS_THRESHOLD denotes the size of the range for that a linear search is faster than using a
+  // binary search on a sorted vector. The value of 200 has been found by a set of simple TPC-H measurements (#2038).
+  static constexpr auto LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD = 200.0f;
+
+  // This methods estimates the threshold up to which is linear search is faster. Since the end position vector is
+  // (depending on actual data) a highly compressed vector of offsets, the offset distance up to which is linear search
+  // is beneficial can be much larger than LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD.
+  static ChunkOffset determine_linear_search_offset_distance_threshold(
+      const std::shared_ptr<const pmr_vector<ChunkOffset>>& end_positions) {
+    if (end_positions->empty()) {
+      return 0;
+    }
+
+    const ChunkOffset chunk_size = end_positions->back();
+    const size_t run_count = end_positions->size();
+
+    const auto avg_elements_per_run = static_cast<float>(chunk_size) / run_count;
+    return static_cast<ChunkOffset>(LINEAR_SEARCH_VECTOR_DISTANCE_THRESHOLD * std::ceil(avg_elements_per_run));
+  }
+
+  using EndPositionIterator = typename pmr_vector<ChunkOffset>::const_iterator;
+  static EndPositionIterator search_end_positions_for_chunk_offset(
```

Comment:
This interface looks like a bunch of things just dumped in here. Wouldn't it be easier if this was non-static?

Reply:
I know it looks ugly and I'll take another look at it. But I am (i) not a big fan of non-static methods and (ii) the function is used in both iterators. I'd guess accessing members within the method isn't possible?!?

Reply:
Could you explain your position on (i)? I didn't notice (ii), that makes sense.

Reply:
Nothing special, just that reasoning is harder when you access "state" of the object. If it would be a member, we could also modify the state variables directly in the method (e.g., `_prev_chunk_offset`). Because why not, if it's a member, right?
But this would make this method even more hard to understand since it's no longer `const`. Might be my way of understanding code, but I often prefer several ugly parameters over "implicitly imported state variable" (there are probably better terms :D).